cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/cache/DynamicDatasetCacheTest.java

```diff
@@ -185,6 +190,57 @@ protected void testDatasetCache(@Nullable Map<String, TestDataset> datasetMap)
     }
   }
 
+  @Test
+  public void testThatDatasetsStayInTransaction() throws TransactionFailureException {
+    final AtomicInteger hash = new AtomicInteger();
+    Transactions.execute(cache.newTransactionContext(), "foo", new Runnable() {
+      @Override
+      public void run() {
+        try {
+          // this writes the value "x" to row "key"
+          TestDataset ds = cache.getDataset("a", ImmutableMap.of("value", "x"));
+          ds.write();
+          // the dataset will go out of scope; remember the dataset's hashcode for later comparison
+          hash.set(ds.hashCode());
```

Comment:
In this case I know the type of the dataset and that it does not override hashCode(). But sure, I can make the change.